Rules to control spaces and newlines around Sass operators

dryoma opened the pull request:
Implements #20 

Along with borrowing some existing stylelint utils I had to add two complex tools: one for pasring media queries and another for parsing strings for Sass operators.

The media parser, I suppose, is better off as a standalone tool. I'm just not sure whether it should be prepared as some kind of a PostCSS-stuff (like [postcss-value-parser](https://github.com/TrySound/postcss-value-parser)), and whether there are any requirements for a standalone tool to be called PostCSS-<something> at all. So, any thought are more than welcome.

As for operations parser: the more I added to it the more I grew to believe we'll eventually need a real full fledged Sass parser. At least the one for operations. Cause all those helpers do help in simple cases, but in complex ones like now they'll just grow too ugly.

The rules' intended mechanics were already described in the [issue](https://github.com/kristerkari/stylelint-scss/issues/20), and to sum it up:
* `operator-no-unspaced` only allows to **force** whitespace around operators, and on both sides. It is because `$var-10` and `$var - 10` are different things in Sass;
* `operator-no-newline-` can prohibit newlines but don't require them. That is to allow users to place multiple operands on the same line. Just like ESLint's [operator-linebreak](http://eslint.org/docs/rules/operator-linebreak).

-------

**TODO:**
- [x] Resolve interpolation in CSS block comments
- [x] Remove media query parser util, replace with `postcss-media-query-parser`.

jeddy3 commented:
Wow. Awesome job!

> The media parser, I suppose, is better off as a standalone tool. 

Yes please. This has been on our bucket list for a while! At least one core rule proposal that will use it springs to mind.

> I'm just not sure whether it should be prepared as some kind of a PostCSS-stuff

I'm not sure either.

kristerkari commented:
Thanks! I'll try to review this soonish.

dryoma commented:
@jeddy3 got it. Do you maybe have requests for some other features (other than `each` and `walk` methods) for the media parser?

jeddy3 commented:
> Do you have requests for some other features (other than each and walk methods) for the media parser?

Nope, those two would be super!

dryoma commented:
@jeddy3 done: https://github.com/dryoma/postcss-media-query-parser (will later remove it from the code here).

jeddy3 commented:
Awesome, thanks!

kristerkari commented on src/rules/operator-no-unspaced/README.md:
````diff
@@ -0,0 +1,96 @@
+  width: 1  
+    - a;
+}
+```
````
Maybe there should also be some examples like these to show what is accepted:

```scss
a { width: 10 * 1; }
```

```scss
a { width: 10 + 1; }
```

dryoma replied:
Sure, I was so busy dealing with edge cases that usual cases ended up completely discriminated against %)

kristerkari commented on src/rules/operator-no-unspaced/README.md:
````diff
@@ -0,0 +1,96 @@
+```
+
+```scss
+a { width: "10*1"; }} // Inside a string, ignored
````
Extra `}`

dryoma commented:
Added: internal media parsing tool, resolving interpolation in CSS comments; exapmles to readme. And fixed some typos.

Any more comments on the code here? Or shall I merge this?

BTW @jeddy3 do you think stylelint might also need a [comment parser](https://github.com/kristerkari/stylelint-scss/blob/13a73323619ac186f184d1704c1b7ad19c0a411b/src/utils/findCommentsInRaws.js) to be able to check comments that are dropped by postcss at the moment? i.e. should it also declare independence.

kristerkari commented on src/rules/operator-no-unspaced/index.js:
```diff
@@ -0,0 +1,237 @@
+          string: item.value,
+          globalIndex: declarationValueIndex(item),
+          // For Sass variable values some special rules apply
+          isAfterColon: item.prop[0] === "$" ? true : false,
```
This could be just `isAfterColon: item.prop[0] === "$"`, no need for ternary.

kristerkari commented on src/utils/sassValueParser/index.js:
```diff
@@ -0,0 +1,788 @@
+  isAfterColon,
+  callback,
+}) {
+  // console.log("===============================================")
```
Maybe remove this? :)

kristerkari commented on src/utils/sassValueParser/index.js:
```diff
@@ -0,0 +1,788 @@
+  // The early check above helps prevent deep recursion here
+  const isPrecedingOperator_ = isPrecedingOperator(string, index)
+  
+  // console.log(string, index, string[index])
```
This could also be removed.

kristerkari commented:
I added a couple of small comments, but other than that I think that this is good to merge.

There is quite a lot of code with all the needed utils/parsers, but there are also a ton of tests to verify that things work correctly.

kristerkari commented on src/rules/operator-no-unspaced/index.js:
```diff
@@ -0,0 +1,237 @@
+
+  root.walk(item => {
+    // console.log('-----------------------')
+    // console.log(item)
```
Maybe remove these console logs too.

jeddy3 commented:
> BTW @jeddy3 do you think stylelint might also need a comment parser, to be able to check comments that are dropped by postcss at the moment?

We [decided](https://github.com/stylelint/stylelint/pull/1683#issuecomment-234721247) just to document this limitation and wait for `postcss@6`. We already have three `comment-*` rules in core. Adding support for selector and value lists to these rules would be a breaking change. So, we figured we'd keep the limitation consistent across the old and new rules, and wait to fix them all with `postcss@6`.

dryoma commented:
Ok, merging this. And waiting for a heckish flood of bug reports %)

> We decided just to document this limitation and wait for postcss@6
Well, I thought it'd be not that productive to wait for the 6th version (and IF they fix the comments in it), since we have a specific rule for forcing/prohibiting // comments to be inline, and the tool was not that difficult to implement.
